diff --git a/in.h b/in.h
index 2b33361..3e46f6c 100644
--- a/in.h
+++ b/in.h
@@ -1,4 +1,8 @@
-c  in.h   contains i-o stuff
-         parameter(ibig=750000, rdepmin=0.0)
-         common /oi/nn,rlat(ibig),rlon(ibig),rdep(ibig),
-     1   alpha(ibig),beta(ibig),rho(ibig),inout(ibig)
+! in.h - memory usage is ibig*37*4 bytes
+!     integer, parameter :: ibig = 4194304
+      integer, parameter :: ibig = 4507502
+      real, parameter :: rdepmin = 0.
+      common /oi/ rlat, rlon, rdep, alpha, beta, rho, inout, nn, nnl
+      real :: rlat(ibig), rlon(ibig), rdep(ibig), alpha(ibig),
+     $  beta(ibig), rho(ibig)
+      integer :: inout(ibig), nn, nnl
diff --git a/iobin.f b/iobin.f
new file mode 100644
index 0000000..70f9e64
--- /dev/null
+++ b/iobin.f
@@ -0,0 +1,80 @@
+! Binary I/O for SCEC CVM
+
+      subroutine readpts(kerr)
+      implicit none
+      include 'in.h'
+      integer :: kerr, nio, i
+      character(160) :: file_lon, file_lat, file_dep
+      write (0, '(a)') 'SCEC Community Velocity Model'
+      open (1, file='cvms-input', status='old')
+      read (1, *) nn
+      read (1, '(a)') file_lon
+      read (1, '(a)') file_lat
+      read (1, '(a)') file_dep
+      close (1)
+      if (nn > ibig) then
+          write (0, *) 'Error: nn greater than ibig', nn , ibig
+          stop
+      end if
+      inquire (iolength=nio) rlon(1:nn)
+      write (0, '(a)') 'Reading input'
+      open (1, file=file_lon, recl=nio, form='unformatted',
+     $    access='direct', status='old')
+      open (2, file=file_lat, recl=nio, form='unformatted',
+     $    access='direct', status='old')
+      open (3, file=file_dep, recl=nio, form='unformatted',
+     $    access='direct', status='old')
+      read (1, rec=1) rlon(1:nn)
+      read (2, rec=1) rlat(1:nn)
+      read (3, rec=1) rdep(1:nn)
+      close (1)
+      close (2)
+      close (3)
+      write (0, '(a)') 'Sampling velocity model'
+      do i = 1, nn
+          if (rdep(i) .lt. 0) write (0, *)
+     $        'Error: degative depth', i, rlon(i), rlat(i), rdep(i)
+          if (rlon(i)/=rlon(i).or.rlat(i)/=rlat(i).or.rdep(i)/=rdep(i))
+     $        write (0, *) 'Error: NaN', i, rlon(i), rlat(i), rdep(i)
+          rdep(i) = rdep(i) * 3.2808399
+          if (rdep(i) .lt. rdepmin) rdep(i) = rdepmin
+      end do
+      kerr = 0
+      end
+
+      subroutine writepts(kerr)
+      implicit none
+      include 'in.h'
+      integer :: kerr, nio, i
+      character (160) :: file_rho, file_alpha, file_beta
+      inquire (iolength=nio) rho(1:nn)
+      open (1, file='cvms-input', status='old')
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_alpha
+      read (1, '(a)') file_beta
+      close (1)
+      write (0, '(a)') 'Writing output'
+      open (1, file=file_rho, recl=nio, form='unformatted',
+     $    access='direct', status='replace')
+      open (2, file=file_alpha, recl=nio, form='unformatted',
+     $    access='direct', status='replace')
+      open (3, file=file_beta, recl=nio, form='unformatted',
+     $    access='direct', status='replace')
+      write (1, rec=1) rho(1:nn)
+      write (2, rec=1) alpha(1:nn)
+      write (3, rec=1) beta(1:nn)
+      close (1)
+      close (2)
+      close (3)
+      do i = 1, nn
+          if (rho(i)/=rho(i).or.alpha(i)/=alpha(i).or.beta(i)/=beta(i))
+     $        write (0, *) 'Error: NaN', i, rlon(i), rlat(i), rdep(i)
+      end do
+      write (0, '(a)') 'Finished'
+      kerr = 0
+      end
+
diff --git a/iompi.f b/iompi.f
new file mode 100644
index 0000000..a7331dc
--- /dev/null
+++ b/iompi.f
@@ -0,0 +1,112 @@
+! MPI I/O for SCEC CVM
+
+      subroutine readpts (kerr)
+      implicit none
+      include 'in.h'
+      include 'mpif.h'
+      integer (kind=mpi_offset_kind) :: offset, nn8, np8, nnl8
+      integer :: kerr, ip, np, ifh, info, ierr, i
+      character (160) :: file_lon, file_lat, file_dep
+      call mpi_init(ierr)
+      open (1, file='cvms-input', status='old')
+      read (1, *) nn8
+      read (1, '(a)') file_lon
+      read (1, '(a)') file_lat
+      read (1, '(a)') file_dep
+      close (1)
+      call mpi_comm_rank(mpi_comm_world, ip, ierr)
+      call mpi_comm_size(mpi_comm_world, np, ierr)
+      call mpi_file_set_errhandler(mpi_file_null,
+     $    mpi_errors_are_fatal, ierr)
+      if (ip == 0) write (0, '(a)') 'SCEC Community Velocity Model'
+      nnl = nn8 / np
+      np8 = np
+      if (modulo(nn8, np8) /= 0) nnl = nnl + 1
+      nnl8 = nnl
+      nn = min(nnl8, nn8 - nnl8 * ip)
+      if (nn > ibig) then
+          write (0, *) 'Error: nn greater than ibig', nn, ibig
+          stop
+      end if
+      call mpi_type_size(mpi_real, i, ierr)
+      offset = i
+      offset = offset * nnl * ip
+      if (ip == 0) write (0, '(a)') 'Reading input'
+      i =  mpi_mode_rdonly
+      info = mpi_info_null
+      !call mpi_info_create(info, ierr)
+      !call mpi_info_set(info, 'romio_cb_read', 'enable', ierr)
+      !call mpi_info_set(info, 'romio_ds_read', 'disable', ierr)
+      !call mpi_file_open(mpi_comm_world, file_lon, i, info, ifh, ierr)
+      call mpi_file_read_at_all(ifh, offset, rlon, nn, mpi_real,
+     $    mpi_status_ignore, ierr)
+      call mpi_file_close(ifh, ierr)
+      call mpi_file_open(mpi_comm_world, file_lat, i, info, ifh, ierr)
+      call mpi_file_read_at_all(ifh, offset, rlat, nn, mpi_real,
+     $    mpi_status_ignore, ierr)
+      call mpi_file_close(ifh, ierr)
+      call mpi_file_open(mpi_comm_world, file_dep, i, info, ifh, ierr)
+      call mpi_file_read_at_all(ifh, offset, rdep, nn, mpi_real,
+     $    mpi_status_ignore, ierr)
+      call mpi_file_close(ifh, ierr)
+      !call mpi_info_free(info, ierr)
+      if (ip == 0) write (0, '(a)') 'Sampling velocity model'
+      do i = 1, nn
+          if (rdep(i).lt.0) write(0, *)
+     $        'Error: negative depth', i, rlon(i), rlat(i), rdep(i)
+          if (rlon(i)/=rlon(i).or.rlat(i)/=rlat(i).or.rdep(i)/=rdep(i))
+     $        write (0, *) 'Error: NaN', i, rlon(i), rlat(i), rdep(i)
+          rdep(i) = rdep(i) * 3.2808399
+          if (rdep(i) .lt. rdepmin) rdep(i) = rdepmin
+      end do
+      kerr = 0
+      end
+
+      subroutine writepts(kerr)
+      implicit none
+      include 'in.h'
+      include 'mpif.h'
+      integer (kind=mpi_offset_kind) :: offset
+      integer :: kerr, ip, ifh, info, ierr, i
+      character (160) :: file_rho, file_alpha, file_beta
+      open (1, file='cvms-input', status='old')
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_rho
+      read (1, '(a)') file_alpha
+      read (1, '(a)') file_beta
+      close (1)
+      call mpi_comm_rank(mpi_comm_world, ip, ierr)
+      call mpi_type_size(mpi_real, i, ierr)
+      offset = i
+      offset = offset * nnl * ip
+      if (ip == 0) write (0, '(a)') 'Writing output'
+      i = mpi_mode_create + mpi_mode_wronly
+      info = mpi_info_null
+      !call mpi_info_create(info, ierr)
+      !call mpi_info_set(info, 'romio_cb_write', 'enable', ierr)
+      !call mpi_info_set(info, 'romio_ds_write', 'disable', ierr)
+      !call mpi_file_open(mpi_comm_world, file_rho, i, info, ifh, ierr)
+      call mpi_file_write_at_all(ifh, offset, rho, nn, mpi_real,
+     $    mpi_status_ignore, ierr)
+      call mpi_file_close(ifh, ierr)
+      call mpi_file_open(mpi_comm_world, file_alpha, i, info, ifh, ierr)
+      call mpi_file_write_at_all(ifh, offset, alpha, nn, mpi_real,
+     $    mpi_status_ignore, ierr)
+      call mpi_file_close(ifh, ierr)
+      call mpi_file_open(mpi_comm_world, file_beta, i, info, ifh, ierr)
+      call mpi_file_write_at_all(ifh, offset, beta, nn, mpi_real,
+     $    mpi_status_ignore, ierr)
+      call mpi_file_close(ifh, ierr)
+      !call mpi_info_free(info, ierr)
+      kerr = 0
+      do i = 1, nn
+          if (rho(i)/=rho(i).or.alpha(i)/=alpha(i).or.beta(i)/=beta(i))
+     $        write (0, *) 'Error: NaN', i, rlon(i), rlat(i), rdep(i)
+      end do
+      if (ip == 0) write (0, '(a)') 'Finished'
+      call mpi_finalize(ierr)
+      end
+
diff --git a/iotxt.f b/iotxt.f
new file mode 100644
index 0000000..2c00fd2
--- /dev/null
+++ b/iotxt.f
@@ -0,0 +1,38 @@
+! ASCII I/O for SCEC CVM
+
+      subroutine readpts(kerr)
+      implicit none
+      include 'in.h'
+      integer :: kerr, i
+      write (0, '(a)') 'SCEC Community Velocity Model'
+      read (*, *) nn
+      if (nn > ibig) then
+          print *, 'Error: nn greater than ibig', nn, ibig
+          stop
+      end if
+      do i = 1, nn
+          read (*, *) rlat(i), rlon(i), rdep(i)
+          if (rdep(i) .lt. 0) write (0, *)
+     $        'Error: degative depth', i, rlon(i), rlat(i), rdep(i)
+          if (rlon(i)/=rlon(i).or.rlat(i)/=rlat(i).or.rdep(i)/=rdep(i))
+     $        write (0, *) 'Error: NaN', i, rlon(i), rlat(i), rdep(i)
+          rdep(i) = rdep(i) * 3.2808399
+          if (rdep(i) .lt. rdepmin) rdep(i) = rdepmin
+      end do
+      kerr = 0
+      end
+
+      subroutine writepts(kerr)
+      implicit none
+      include 'in.h'
+      integer :: kerr, i
+      do i = 1, nn
+          rdep(i) = rdep(i) / 3.2808399
+          if (rho(i)/=rho(i).or.alpha(i)/=alpha(i).or.beta(i)/=beta(i))
+     $        write (0, *) 'Error: NaN', i, rlon(i), rlat(i), rdep(i)
+          write (*, '(f8.5,1x,f10.5,1x,f9.2,1x,f8.1,1x,f8.1,1x,f8.1)')
+     $        rlat(i), rlon(i), rdep(i), alpha(i), beta(i), rho(i)
+      end do
+      kerr = 0
+      end
+
diff --git a/names.h b/names.h
index 3fbaa93..f098902 100644
--- a/names.h
+++ b/names.h
@@ -1,5 +1,5 @@
 c names.h   name roots for edges and surfaces
-         character*4 aedname(numsur)
+         character(4) aedname(numsur)
          data (aedname(i),i=1,numsur)/'b1__','b2__','b3__','b4__',
      1    'b5__','ku1_','ku2_',
      2    'ku3_','ku4_','ku5_','ku8_','pu1_','pu2A','pu2B','pu3_',
diff --git a/version2.2.f b/version2.2.f
index 0e004f6..5c3bb27 100644
--- a/version2.2.f
+++ b/version2.2.f
@@ -19,6 +19,7 @@ c scum v2e     4-00   added variable Moho. HM.
 c scum v2f,g   6-00   modified geotech stuff. HM
 c scum v2h     6-00   modified geotech , separate P and S. HM
 c scum v2i     8-00   new tomo interpolator, vent glitch fixes HM
+c Bug fixes and modifications for binary and MPI I/O. Ely 2010/11/23
 c                   
 c
          include 'in.h'
@@ -33,6 +34,36 @@ c
          include 'surfaced.h'
          include 'genpro.h'
          include 'genprod.h'
+
+c initialize to zero. Ely
+      inct = 0
+      incto = 0
+      iupm = 0
+      idnm = 0
+      rshcor = 0.0
+      do i = 1,isurmx
+        iiiold(i) = 0
+        inorold(i) = 0
+        rsuqold(i) = 0.0
+      end do
+      rtemp01 = 0.0
+      rtemp05 = 0.0
+      rtemp07 = 0.0
+      rtemp22 = 0.0
+      rtemp36 = 0.0
+      rtemp47 = 0.0
+      rtemp50 = 0.0
+      rtemp55 = 0.0
+      rtemp56 = 0.0
+      rtemp57 = 0.0
+      rtemp62 = 0.0
+      rtemp63 = 0.0
+      rtemp64 = 0.0
+      rtemp65 = 0.0
+      rtemp68 = 0.0
+      rtemp69 = 0.0
+      rtemp70 = 0.0
+
 c some constants
          rd2rad=3.141593/180.
          rckval=500000.
@@ -511,7 +542,7 @@ c check for LAB and SMM and SAN BERDO ---99 is a flag---
 c--diag-----write(*,*)l0,iup,idn
             go to 1177
             endif
-1188       iup=idn
+           iup=idn
            iupm=idnm
            ishal=idn
            goto 1177
@@ -848,7 +879,7 @@ c------------- write out points and values---------------
 c-----reads strat surface edges---------------
 c---reads x-y pairs--------------------------
          include 'surface.h'
-         character*9 aname2, asuf2*5
+         character(9) aname2, asuf2*5
          include 'names.h'
          asuf2='_edge'
          kerr=0
@@ -871,7 +902,7 @@ c-------------------------------
 c-----read model edge file, kept separate from surface edge-----
 c-------to simplify index counts--------------------------------
           include 'surface.h'
-          character*9 aname3
+          character(9) aname3
           kerr=0
 c---file name assignment-------
           aname3='bmod_edge'
@@ -890,13 +921,13 @@ c        read(18,*)i3   !i3 now set at top
          subroutine readgeo(k2err)
 c reads geologic contours
          include 'sgeo.h'
-         character*16 fileii,a1*10
+         character(16) fileii,a1*10
          include 'sgeod.h'
 c---file name assignment-------
          fileii='lab_geo2_geology'
          k2err=0
 c read file
-           open(12,file=fileii,type='old',err=977)
+           open(12,file=fileii,status='old',err=977)
            do 300 k=1,ngeo
            read(12,*)np(k)
             do 310 k1=1,np(k)
@@ -963,7 +994,7 @@ c
 c---reads stratigraphic surfaces-------------------------
 c---reads spyglass ascii output--------------------------
          include 'surface.h'
-         character*8 aname, asuf*4
+         character(8) aname, asuf*4
          include 'names.h'
          asuf='_sur'
          kerr=0
@@ -1004,7 +1035,7 @@ c---For SGV (i=72,73,74) convert depth thousands of feet to feet---
 c-reads uplift file for LAB--------------------
 c-----just like readsur-------------
         include 'labup.h'
-        character*8 filei4
+        character(8) filei4
 c file name assignment
         filei4='laup_sur'
 c
@@ -1013,6 +1044,8 @@ c
          do 400 i=1,nlaup 
           read(16,*) rlaup(i),(rzupl(i,j),j=1,nloup) 
           do 440 j=1,nloup
+c rloup is not set anywhere, at least initialize it to zero
+          rloup(j) = 0.0
 c convert thousands of feet to feet---------------
           rzupl(i,j)=rzupl(i,j)*1000. 
 440       continue 
@@ -1027,7 +1060,7 @@ c convert thousands of feet to feet---------------
 c--gets uplift for LAB---------------
 c find uplift amount at current lat long
         include 'labup.h'
-877        do 817 l7=1,nlaup-1
+           do 817 l7=1,nlaup-1
            if(rlatl0.le.rlaup(l7).and.rlatl0.gt.rlaup(l7+1))then
             do 828 l8=1,nloup-1 
             if(rlonl0.gt.rloup(l8).and.rlonl0.le.rloup(l8+1))then
@@ -1131,45 +1164,11 @@ c
          return
          end
 
-         subroutine readpts(kerr)
-c-----read points of interest-----------------
-         include 'in.h'
-         kerr=0
-         open(15,file='btestin',status='old',err=1099) 
-c        nn=51456
-         read(15,*)nn 
-         do 150 i=1,nn 
-         read(15,*)rlat(i),rlon(i),rdep(i) 
-c now read in meters
-         rdep(i)=rdep(i)*3.2808399
-         if(rdep(i).lt.rdepmin)rdep(i)=rdepmin 
-150      continue 
-         close(15)
-         go to 1088
-1099     kerr=1
-1088     return
-         end
- 
-         subroutine writepts(kerr)
-c----write points of interest-----------------
-         include 'in.h'
-         kerr=0
-         open(17,file='btestout',status='new')
-         do 155 i=1,nn
-          rdep(i)=rdep(i)/3.2808399
-          write(17,77)rlat(i),rlon(i),rdep(i),alpha(i)
-     1    ,beta(i),rho(i)
-77       format(f8.5,1x,f10.5,1x,f9.2,1x,f8.1,1x,f8.1,1x,f8.1)
-155      continue
-         close(17)
-         return
-         end
-
          subroutine readivsurf(kerr)
 c-----read Imperial Valley surfaces--------------
          include 'in.h'
          include 'ivsurface.h'
-         character*8 aname4, asuf4*6,asrnam(numsiv)*2
+         character(8) aname4, asuf4*6,asrnam(numsiv)*2
          data (asrnam(i),i=1,numsiv)/'25','55','60','65',
      1    '70','Mo'/
          asuf4='.ascii'
@@ -1253,6 +1252,7 @@ c--Calculates the Imperial Valley model velocities--
          include 'ivsurfaced.h'
          include 'generic_loc.h'
 c---see if in constrained or generic Imperial Valley
+         rd2rad = 3.141593 / 180.0
          xref=rmoivx(ivi3)
          yref=rmoivy(ivi3)
          reflat=yref*rd2rad
@@ -1352,9 +1352,14 @@ c -- define the regional tomo velocities -----------------------
          include 'regional.h'
          include 'regionald.h'
          dimension vervep(4),verves(4)
+         rscal = 0.0
+         iinum = 0
          rd2rad=3.141593/180.
 c -- find which box point is in--
          do 1927 n=1,nregll-ninrow
+          do 8181 j=1,4
+          nearn(j)=0
+8181      continue
          rckbox=mod(n,ninrow)
 c -- avoid trying to make box off row ends
          if(rckbox.eq.0.)go to 1927
@@ -1377,9 +1382,6 @@ c -- prepare for inside
           x22(i) = (rboxlo(i) - xref)*111.1*cos(reflat)
 8179     continue
 c -- see if inside this box
-          do 8181 j=1,4
-          nearn(j)=0
-8181      continue
           yp=(rla-yref)*111.1
           xp=(rlo-xref)*111.1*cos(reflat)
           call inside(xp,yp,x22,y22,nbox,ins)
@@ -1533,15 +1535,15 @@ c Do the interpolation
       return
       end
 
-         subroutine readbore(kerr) 
+         subroutine readbore(k2err)
 c--read geotech borehole data-------------- 
          include 'borehole.h'
-         character*9 fileib
+         character(9) fileib
 c---file name assignment-------
          fileib='boreholes'
          k2err=0
 c read file
-         open(15,file=fileib,type='old',err=2978)
+         open(15,file=fileib,status='old',err=2978)
          iprono=0
          ibhct=0
          ieach=0
@@ -1572,15 +1574,15 @@ c read file
 2915      return
            end
 
-         subroutine readgene(kerr)
+         subroutine readgene(k2err)
 c--read generic borehole profiles--------------
          include 'genpro.h'
-         character*12 fileig,ag1*50
+         character(12) fileig,ag1*50
 c---file name assignment-------
          fileig='soil_generic'
          k2err=0
 c read file
-         open(12,file=fileig,type='old',err=2977)
+         open(12,file=fileig,status='old',err=2977)
          do 2300 k=1,numgen
           read(12,*)irt2
           numptge2(k)=irt2
@@ -1749,11 +1751,12 @@ c chino - berdo need to split this!
          return
         end
 
-        subroutine readsoil(kerr)
+        subroutine readsoil(k2err)
 c--reads soil type info---------------------------------
 c Reads a modified .pgm ascii file
         include 'soil1.h'
-        character*50 filesb
+        character(50) filesb
+        k2err = 0
 c here's input file name-----------------------------------
         filesb='soil2.pgm'
         open(16,file=filesb,status='old',err=5977)
@@ -1786,12 +1789,17 @@ c iradcts=number of nearby boreholes with data
          include 'genpro.h'
          include 'wtbh1.h'
          include 'wtbh2.h'
+         roff = 0.0
+         rvte3 = 0.0
+         rvte8 = 0.0
          ihtfg=0
          rtvelges=0.
          rdep=rdep2
          do 7013 n=1,nrad
          iradcts(n)=0
          radvs(n)=0.
+         rtvels(n) = 0.0
+         rtewts(n) = 0.0
 7013     continue
 c--check ifs flag--
          if(ifs.eq.0)then
@@ -1915,9 +1923,12 @@ c iradctp=number of nearby boreholes with data
          include 'wtbh1.h'
          include 'wtbh3.h'
          rtvelges=0.
+         rva = 0.0
          do 97013 n=1,nrad
          iradctp(n)=0
          radvp(n)=0.
+         rtvelp(n) = 0.0
+         rtewtp(n) = 0.0
 97013     continue
           do 97011 l=2,nrad
            do 97012 i=1,iradct(l)
@@ -1993,10 +2004,11 @@ c remember a better number to interpolate from rather than hard generic
          return
          end
 
-        subroutine readmoho(kerr)
+        subroutine readmoho(k2err)
 c--reads the moho surface file--------------------------
         include 'moho1.h'
-        character*50 filemo
+        character(50) filemo
+        k2err = 0
 c here's input file name-----------------------------------
         filemo='moho_sur'
         open(16,file=filemo,status='old',err=3233)
